### ISSUE TITLE
[lit-GPT] Thunder with torch.compile executor performs consistently worse than Thunder on all model sizes/batch sizes on Pythia models

### ISSUE THREAD

**Issue by parthmannan** [Lightning-AI/lightning-thunder#256]
## 🐛 Bug

The performance of using the hybridized torch.compile executor w/ Thunder is worse than plain Thunder on Pythia models. These set of models differ from LLaMa architecture in few main ways - 

1. Use LayerNorm instead of RMSNorm
2. Use GeLU instead of 'SiLU(x) * x`
3. Uses parallel residual (i.e. the MLP block is computed with an input computed before the Attention block, not after)

Example performance on H100 Single Node FP16 for Pythia6.9B, MBS=1, GBS=8, FSDP ZeRO2 w/o bucketing
Thunder iteration time (ms) = 232.74 ms
Thunder + torch.compile iteration time (ms) = 239.23 ms

cc @crcrpar @apaz-cli

**Comment by IvanYashchuk** [comment]
Checking the performance today I see the following:
Thunder+TorchCompileCat: 247.54 ms
Thunder: 251.69 ms
```
torchrun --nproc_per_node=8 thunder/benchmarks/benchmark_litgpt.py --compile=thunder_inductor_cat --distributed_mode=fsdp --nsys_enabled=False --micro_batch_size=1 --global_batch_size=8 --model_name=pythia-6.9b
torchrun --nproc_per_node=8 thunder/benchmarks/benchmark_litgpt.py --compile=thunder --distributed_mode=fsdp --nsys_enabled=False --micro_batch_size=1 --global_batch_size=8 --model_name=pythia-6.9b
```
The numbers are worse than in Feb, but TorchCompileCat brings some value here, but that could be also due to regressions in fsdp. The major difference, I think, from Feb is that nvFuser now fuses `cat` since https://github.com/Lightning-AI/lightning-thunder/pull/35.

Here are the numbers for single GPU execution, TorchCompileCat still improves performance a bit:
Thunder+TorchCompileCat: 208.25 ms
Thunder: 211.72 ms

```
python thunder/benchmarks/benchmark_litgpt.py --compile=thunder_inductor_cat --micro_batch_size=1 --model_name=pythia-6.9b
python thunder/benchmarks/benchmark_litgpt.py --compile=thunder --micro_batch_size=1 --model_name=pythia-6.9b
```

TorchCompileCat is a hack for executing RoPE fusions, it's by accident that this executor also claims backward of `torch.split` (that is a `cat` op) and consequently breaks nvFuser fusions. TorchCompileCat should be further constrained to be applied just for RoPE. And we should reevaluate its performance on a wider range of models and microbatch sizes.

Let's try constraining it a bit:
```diff
diff --git a/thunder/executors/torch_compile.py b/thunder/executors/torch_compile.py
index 1e18c42e..3f20136c 100644
--- a/thunder/executors/torch_compile.py
+++ b/thunder/executors/torch_compile.py
@@ -198,10 +198,10 @@ from thunder.executors.torchex import ex as pytorch_ex
 # since they would be competing over fusion opportunities. The advantage over simply doing `torch.compile` is that you
 # still get all of Thunder's advantages, like enabling custom executors (e.g. with custom triton kernels) before it.
 required_ops = {
-    "torch.cat",
+    #"torch.cat",
     prims.cat.id,
-    prims.pad.id,
-    prims.slice_prim.id,
+    #prims.pad.id,
+    #prims.slice_prim.id,
 }
 torch_compile_cat_ex = TorchCompileExecutor(name="torchcompile_cat", required_ops=required_ops)
 register_executor(torch_compile_cat_ex)
```

Here's how the forward trace is affected:
1. `phi_a` is not saved for backward anymore
https://github.com/Lightning-AI/lightning-thunder/blob/3034ef9efad3db43f9e94e3e9013ce853c0b3680/thunder/torch/__init__.py#L1426

Here's how the backward trace is affected:
1. There's no TorchCompile fusion anymore, all backward fusions are using nvFuser
2. nvFusion1 was previously interrupted and its output was consumed only by interrupting TorchCompile region, now it returns one of the grads and intermediate consumed by matmul
3. nvFusion2 now merges the work done by the the TorchCompile0 and nvFusion2 regions before the change

Let's check how this change impacts our current microbenchmark (which should be revised and updated!):
```py
pytest thunder/benchmarks/targets.py -k "test_llama2_qkv_split_rope_7b_train[thunder+nvfuser+torch.compile]"
```
The results are worse with this change on Llama 2 7B arch: 8.6 ms vs 19.5 ms, and for Pythia 6.9B it's 4.4 ms vs 9.8 ms. Running Pythia 6.9B on a single GPU with or without this change doesn't have an impact on perf though.

**Comment by parthmannan** [comment]
That's quite interesting.

>The major difference, I think, from Feb is that nvFuser now fuses cat since https://github.com/Lightning-AI/lightning-thunder/pull/35.

Do we know if nvFuser enabling cat is just functionally good or performant as well?
I believe TorchCompileCat executor was added for concats specifically so using that without `cat` is probably expected to be worse if the nvFuser support isn't as performant.

I can look into that. But even if `thunder_inductor_cat` improves perf a little bit, we may be leaving perf on the table as it improves `RoPE` perf but degrades other aspects of the trace so there should still be similar room for improvement.

**Comment by parthmannan** [comment]
I also just dived into the trace again to see if the behavior is the same but it has changed. And this is because the Thunder trace itself has changed a little bit.
Rolling back to a previous comment (see above for the calculation) - 
```
dGeLU(x) = A + (B * C) / f337
```

Earlier we had, nvFusion compute B, C separately. TorchCompile compute B*C and then nvFuser compute the final output.

Now we have, nvFusion computes B, C and B*C in a single region and TorchCompile only does A + B*C and has the final output. Somehow, with the change in the Thunder trace - we don't see the same breaks in the nvFuser regions. This is why the performance on Pythia looks decent now. This is quite interesting and I am thinking I have to re-analyze other networks like Phi and Dolly and confirm whether the TorchCompile cat executor is still a perf issue there.

**Comment by IvanYashchuk** [comment]
> Do we know if nvFuser enabling cat is just functionally good or performant as well?

I don't know it and I've created https://github.com/Lightning-AI/lightning-thunder/pull/479 to simplify answering this question.

**Comment by IvanYashchuk** [comment]
@kiya00, could you please help identify what are the best executor options (nvFuser/Inductor/Apex) separately for the forward and backward of this region on the range of all model configurations from LitGPT?
The relevant benchmark in question is this one https://github.com/Lightning-AI/lightning-thunder/blob/d1b016a58a48e5c6282622de488be8c9135dd821/thunder/benchmarks/targets.py#L535
```
pytest thunder/benchmarks/targets.py -k "test_litgpt_qkv_split_rope" --benchmark-group-by='param:config,param:bs,param:compute_type'
```
there's also an environment variable to launch more benchmarks https://github.com/Lightning-AI/lightning-thunder/blob/d1b016a58a48e5c6282622de488be8c9135dd821/thunder/benchmarks/targets.py#L54

**Comment by IvanYashchuk** [comment]
> I'm not very clear about the background of this issue, should this benchmark have better perf using thunder+nvfuser+torch.compile at least on Pythia?

Thunder (and thunder+nvfuser+torch.compile) should have better performance for all cases. The purpose of these benchmarks is to evaluate the current situation and identify what needs to be done to improve performance for worst performing cases.
Besides the logs it would be useful to have a script that analyzes the json results from pytest-benchmark and creates a summary:
* What is the top executor for batchsize 1 and 2 among all configs
* What is the percentage of the configs where an executor is the best for each executor
* What is the max/min/mean/median gap between top executor and Thunder options

Include summary of any other important information that is the json files. A table could be useful, something like:

| Metric                                  | Batch Size 1 | Batch Size 2 |
|-----------------------------------------|--------------|--------------|
| Top Executor                            | Executor A   | Executor B   |
| Percentage of Configs Best for Executor |              |              |
| - Executor A                            | 60%          | 50%          |
| - Executor B                            | 30%          | 40%          |
| - Executor C                            | 10%          | 10%          |
| Gap Between Top Executor and Thunder    |              |              |
| - Max Gap                               | 15 ms        | 20 ms        |
| - Min Gap                               | 1 ms         | 2 ms         |
| - Mean Gap                              | 8 ms         | 10 ms        |
| - Median Gap                            | 7 ms         | 9 ms         |

If the batchsize is increased how do the results change? Maybe the overheads are too large for Thunder? How does the patch from https://github.com/Lightning-AI/lightning-thunder/issues/256#issuecomment-2136096575 affects numbers of `thunder+nvfuser+torch.compile` executor?

Maybe it's possible to get pure CUDA kernel times with a timer from nvFuser https://github.com/NVIDIA/Fuser/blob/18750278f9f20a817808dc1c63c0fb6962d37c9c/benchmarks/python/core.py#L209-L229.

**Comment by kiya00** [comment]
[issue256_analysis.xlsx](https://github.com/user-attachments/files/15968760/issue256_analysis.xlsx)
Here is some initial information for reference (based on container 0614)

**Comment by kiya00** [comment]
> This is quite interesting and I am thinking I have to re-analyze other networks like Phi and Dolly and confirm whether the TorchCompile cat executor is still a perf issue there.

Hi @parthmannan , checking the performance today for the original problem with commands(on H100, container0626)
```
torchrun --nproc_per_node=8 thunder/benchmarks/benchmark_litgpt.py --compile=thunder_inductor_cat --distributed_mode=fsdp --nsys_enabled=False --micro_batch_size=1 --global_batch_size=8 --model_name=pythia-6.9b
torchrun --nproc_per_node=8 thunder/benchmarks/benchmark_litgpt.py --compile=thunder --distributed_mode=fsdp --nsys_enabled=False --micro_batch_size=1 --global_batch_size=8 --model_name=pythia-6.9b
python thunder/benchmarks/benchmark_litgpt.py --compile=thunder_inductor_cat --micro_batch_size=1 --model_name=pythia-6.9b
python thunder/benchmarks/benchmark_litgpt.py --compile=thunder --micro_batch_size=1 --model_name=pythia-6.9b
```
| pythia-6.9b          | zero2     | single gpu |
| -------------------- | --------- | ---------- |
| thunder_inductor_cat | 234.84 ms | 208.45 ms  |
| thunder              | 233.02 ms | 210.97 ms  |

| phi-2                | zero2     | single gpu |
| -------------------- | --------- | ---------- |
| thunder_inductor_cat | 114.54 ms | 104.35 ms  |
| thunder              | 113.44 ms | 105.14 ms  |

| dolly-v2-7b          | zero2     | single gpu |
| -------------------- | --------- | ---------- |
| thunder_inductor_cat | 233.81 ms | 208.45 ms  |
| thunder              | 233.26 ms | 211.03 ms  |

The performance of thunder_inductor_cat exe seems to be decent now. I'll dig into the reason why the RoPE microbenchmark is not what we expected, but for the original problem in this issue, do we want more analysis on that?

cc: @IvanYashchuk

**Comment by IvanYashchuk** [comment]
It's great that "thunder_inductor_cat" is now better for Phi-2 and Dolly, thank you for rerunning the benchmarks! I'm inclined to close this particular issue and start a new one specifically for RoPE microbenchmark performance. We need to understand better how impactful improvements for that microbenchmark for full network runs.

**Comment by parthmannan** [comment]
Given this data, we can definitely close this issue. Thanks so much @kiya00 for re-running this and the analysis.
Just curious, do we know what changed in the partitioning logic that the performance issues are gone now?

**Comment by IvanYashchuk** [comment]
No, I don't know what was changed there.
